clang/lib/Driver/Driver.cpp

```diff
@@ -4995,14 +4995,33 @@ void Driver::BuildActions(Compilation &C, DerivedArgList &Args,
 
   // For an FPGA archive, we add the unbundling step above to take care of
   // the device side, but also unbundle here to extract the host side
-  for (const auto &LI : LinkerInputs) {
+  for (auto &LI : LinkerInputs) {
     Action *UnbundlerInput = nullptr;
+    auto wrapObject = [&]() {
+      bool EarlyLink = false;
+      if (const Arg *A = Args.getLastArg(options::OPT_fsycl_link_EQ))
+        EarlyLink = A->getValue() == StringRef("early");
+      if (Args.hasArg(options::OPT_fintelfpga) && EarlyLink) {
+        // Only wrap the object with -fsycl-link=early
+        auto *BC = C.MakeAction<OffloadWrapperJobAction>(LI, types::TY_LLVM_BC);
+        auto *ASM = C.MakeAction<BackendJobAction>(BC, types::TY_PP_Asm);
+        LI = C.MakeAction<AssembleJobAction>(ASM, types::TY_Object);
+      }
+    };
     if (auto *IA = dyn_cast<InputAction>(LI)) {
       if (IA->getType() == types::TY_FPGA_AOCR ||
           IA->getType() == types::TY_FPGA_AOCX) {
         // Add to unbundler.
         UnbundlerInput = LI;
+      } else {
+        std::string FileName = IA->getInputArg().getAsString(Args);
+        if ((IA->getType() == types::TY_Object && !isObjectFile(FileName)) ||
```

Comment:
How is this different from:
```suggestion
        if (!IA->getType() == types::TY_Object ||
```
or
```suggestion
        if (!isObjectFile(FileName) ||
```

Reply:
Use of `isObjectFile` is based on file extension and doesn't check the type explicitly.  It also considers `-lfile` as objects.  It's a catch-all for items that should be considered for the linker.  We need to also check the type in these instances so we can differentiate from a known archive that we don't want to be considered an 'object'